### DISCUSSION TITLE
Extrapolation in rcjk components

### DISCUSSION THREAD

**Discussion by punchcutter** [fontra/fontra#306] [General]
I'm not sure the best place to put this, but since I ran into it with fontra I'll start here. I have some old source files converted to rcjk where some deep components go outside the axis range. In this case 0-1000 range and here and there we have -500 or 1500 which worked fine in the original (proprietary old and now dead) app as well as when converting to Glyphs. But when I open in fontra anything outside the axis range is clamped so here I get 0 and 1000 instead of -500 and 1500. I guess the bigger question is if that's eventually going to work with https://github.com/harfbuzz/boring-expansion-spec/blob/main/glyf1-varComposites.md
I'm not a fan of the sources I have so if this is never going to work then I'm happy to drop it and start redoing the source components, but wanted to see what everyone thinks first.

**Comment by justvanrossum** [comment]
For better or for worse, Fontra tries to stay close to how variations work in OT. We may want to revisit this, as extrapolation is generally useful during design.

I doubt whether extrapolation is desirable at the OT level: it doesn't allow it for "normal" variable glyphs, and I don't see why variable components should be treated differently. Also: the variations in variable components in `glyf1` use the same (`gvar`) mechanism as outlines, so are (currently) bound to the same limitations.

**Reply by punchcutter** [reply]
That's what I expected, but thought I'd start the discussion anyway just to see. In my particular case these extrapolations would feed directly into VarC so directly into `glyf1` which is why I thought it's probably an issue. If extrapolation _were_ allowed in fontra (well, the underlying variation models) then it seems there would still need to be some way to only allow it for some things and error out if trying to actually build a future `VarC-glyf1-avar2-etc` font.

**Comment by justvanrossum** [comment]
Btw. a somewhat tangential FYI in case you didn't already see: Fontra also supports this UFO extension: https://github.com/googlefonts/variable-components-in-ufo (But it hasn't gotten a lot of miles yet, so it may be buggy.)

**Reply by punchcutter** [reply]
I thought that's what `rckjtools.project.RoboCJKProject.saveVarCoUFO` does, but seems a little different from a quick look.

**Reply by justvanrossum** [reply]
Yeah, that's something older and different, and was specifically meant as an intermediate format to produce a VF with variable components using our `VarC` proposal (on which part of `glyf-1` is based).

**Comment by davelab6** [comment]
I wasn't aware of https://github.com/fonttools/fonttools/pull/2757  and I'm so happy to see it :)
